core/embed/rust/src/ui/layout/util.rs

```diff
@@ -32,3 +40,177 @@ where
     // Returns error if array.len() != N
     vec.into_array().map_err(|_| err)
 }
+
+fn hexlify<'a>(data: &[u8], buffer: &'a mut [u8]) -> &'a str {
+    const HEX_LOWER: [u8; 16] = *b"0123456789abcdef";
+    let mut i: usize = 0;
+    for b in data.iter().take(buffer.len() / 2) {
+        let hi: usize = ((b & 0xf0) >> 4).into();
+        let lo: usize = (b & 0x0f).into();
+        buffer[i] = HEX_LOWER[hi];
+        buffer[i + 1] = HEX_LOWER[lo];
+        i += 2;
+    }
+    // SAFETY: only <0x7f bytes are used to construct the string
+    unsafe { str::from_utf8_unchecked(&buffer[0..i]) }
+}
+
+pub enum StrOrBytes {
+    Str(StrBuffer),
+    Bytes(Obj),
+}
+
+impl StrOrBytes {
+    pub fn as_str_offset<'a>(&'a self, offset: usize, buffer: &'a mut [u8]) -> &'a str {
+        match self {
+            StrOrBytes::Str(x) => &x.as_ref()[offset..],
+            StrOrBytes::Bytes(x) => Self::hexlify(*x, offset, buffer),
+        }
+    }
+
+    fn hexlify(obj: Obj, offset: usize, buffer: &mut [u8]) -> &str {
+        if !obj.is_type_bytes() {
+            return "ERROR";
+        }
+
+        // Convert offset to byte representation, handle case where it points in the
+        // middle of a byte.
+        let bin_off = offset / 2;
+        let hex_off = offset % 2;
+
+        // SAFETY:
+        // (a) only immutable references are taken
+        // (b) reference is discarded before returning to micropython
+        let bin_slice = if let Ok(buf) = unsafe { get_buffer(obj) } {
+            &buf[bin_off..]
+        } else {
+            return "ERROR";
+        };
+        let hexadecimal = hexlify(bin_slice, buffer);
+
+        &hexadecimal[hex_off..]
+    }
+}
+
+impl TryFrom<Obj> for StrOrBytes {
+    type Error = Error;
+
+    fn try_from(obj: Obj) -> Result<Self, Error> {
+        if obj.is_str() {
+            Ok(StrOrBytes::Str(obj.try_into()?))
+        } else if obj.is_type_bytes() {
```

Comment:
we should probably add `obj.is_bytes()` that will check if `obj` is one of `bytes`, `bytearray` or `memoryview`

Reply:
ccfd8ee9f